File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/dialogs/GenericInputDialog.java
/*
 * Copyright (C) 2016  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;

import com.afollestad.materialdialogs.MaterialDialog;

import java.io.Serializable;

public class GenericInputDialog extends DialogFragment implements MaterialDialog.InputCallback {
    private static final String ARG_BUILDER = "builder";
    private static final String ARG_TARGET = "target";
    private static final String ARG_TAG = "tag";

    private Builder mBuilder;
    private DialogListenerTarget mTarget;
    private String mTag;

    public interface GenericInputDialogListener {
        void onConfirmInput(@Nullable String tag, String text);
    }

    @Nullable
    private GenericInputDialogListener getOwner() {
        switch (mTarget) {
        case ACTIVITY:
            return (GenericInputDialogListener) getActivity();
        case FRAGMENT:
            return (GenericInputDialogListener) getTargetFragment();
        case NONE:
        default:
            return null;
        }
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        mBuilder = (Builder) args.getSerializable(ARG_BUILDER);
        mTarget = (DialogListenerTarget) args.getSerializable(ARG_TARGET);
        mTag = args.getString(ARG_TAG);

        MaterialDialog.Builder builder = new MaterialDialog.Builder(getActivity());

        CharSequence hint = null;
        CharSequence prefill = null;

        if (mBuilder.mTitle != null) {
            builder.title(mBuilder.mTitle);
        } else if (mBuilder.mTitleResId != 0) {
            builder.title(mBuilder.mTitleResId);
        }

        if (mBuilder.mMessage != null) {
            builder.content(mBuilder.mMessage);
        } else if (mBuilder.mMessageResId != 0) {
            builder.content(mBuilder.mMessageResId);
        }

        if (mBuilder.mPositive != null) {
            builder.positiveText(mBuilder.mPositive);
        } else if (mBuilder.mPositiveResId != 0) {
            builder.positiveText(mBuilder.mPositiveResId);
        }

        if (mBuilder.mNegative != null) {
            builder.negativeText(mBuilder.mNegative);
        } else if (mBuilder.mNegativeResId != 0) {
            builder.negativeText(mBuilder.mNegativeResId);
        }

        if (mBuilder.mHint != null) {
            hint = mBuilder.mHint;
        } else if (mBuilder.mHintResId != 0) {
            hint = getText(mBuilder.mHintResId);
        }

        if (mBuilder.mPrefill != null) {
            prefill = mBuilder.mPrefill;
        } else if (mBuilder.mPrefillResId != 0) {
            prefill = getText(mBuilder.mPrefillResId);
        }

        if (mBuilder.mInputType != 0) {
            builder.inputType(mBuilder.mInputType);
        }

        builder.input(hint, prefill, mBuilder.mAllowEmpty, this);

        Dialog dialog = builder.build();

        setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);

        return dialog;
    }

    @Override
    public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
        GenericInputDialogListener owner = getOwner();
        if (owner != null) {
            owner.onConfirmInput(mTag, input.toString());
        }
    }

    public static class Builder implements Serializable {
        @Nullable
        String mTitle;
        @StringRes
        int mTitleResId;
        @Nullable
        String mMessage;
        @StringRes
        int mMessageResId;
        @Nullable
        String mPositive;
        @StringRes
        int mPositiveResId;
        @Nullable
        String mNegative;
        @StringRes
        int mNegativeResId;
        @Nullable
        String mHint;
        @StringRes
        int mHintResId;
        @Nullable
        String mPrefill;
        @StringRes
        int mPrefillResId;
        boolean mAllowEmpty = true;
        int mInputType;

        @NonNull
        public Builder title(@Nullable String title) {
            mTitle = title;
            mTitleResId = 0;
            return this;
        }

        @NonNull
        public Builder title(@StringRes int titleResId) {
            mTitle = null;
            mTitleResId = titleResId;
            return this;
        }

        @NonNull
        public Builder message(@Nullable String message) {
            mMessage = message;
            mMessageResId = 0;
            return this;
        }

        @NonNull
        public Builder message(@StringRes int messageResId) {
            mMessage = null;
            mMessageResId = messageResId;
            return this;
        }

        @NonNull
        public Builder positive(@Nullable String text) {
            mPositive = text;
            mPositiveResId = 0;
            return this;
        }

        @NonNull
        public Builder positive(@StringRes int textResId) {
            mPositive = null;
            mPositiveResId = textResId;
            return this;
        }

        @NonNull
        public Builder negative(@Nullable String text) {
            mNegative = text;
            mNegativeResId = 0;
            return this;
        }

        @NonNull
        public Builder negative(@StringRes int textResId) {
            mNegative = null;
            mNegativeResId = textResId;
            return this;
        }

        @NonNull
        public Builder hint(@Nullable String text) {
            mHint = text;
            mHintResId = 0;
            return this;
        }

        @NonNull
        public Builder hint(@StringRes int textResId) {
            mHint = null;
            mHintResId = textResId;
            return this;
        }

        @NonNull
        public Builder prefill(@Nullable String text) {
            mPrefill = text;
            mPrefillResId = 0;
            return this;
        }

        @NonNull
        public Builder prefill(@StringRes int textResId) {
            mPrefill = null;
            mPrefillResId = textResId;
            return this;
        }

        @NonNull
        public Builder allowEmpty(boolean allowEmpty) {
            mAllowEmpty = allowEmpty;
            return this;
        }

        @NonNull
        public Builder inputType(int type) {
            mInputType = type;
            return this;
        }

        @NonNull
        public GenericInputDialog buildFromFragment(@Nullable String tag,
                                                    @NonNull Fragment parent) {
            Bundle args = new Bundle();
            args.putSerializable(ARG_BUILDER, this);
            args.putSerializable(ARG_TARGET, DialogListenerTarget.FRAGMENT);
            args.putString(ARG_TAG, tag);

            GenericInputDialog dialog = new GenericInputDialog();
            dialog.setTargetFragment(parent, 0);
            dialog.setArguments(args);
            return dialog;
        }

        @NonNull
        public GenericInputDialog buildFromActivity(@Nullable String tag) {
            Bundle args = new Bundle();
            args.putSerializable(ARG_BUILDER, this);
            args.putSerializable(ARG_TARGET, DialogListenerTarget.ACTIVITY);
            args.putString(ARG_TAG, tag);

            GenericInputDialog dialog = new GenericInputDialog();
            dialog.setArguments(args);
            return dialog;
        }
    }
}

File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/dialogs/GenericInputDialog.kt
/*
 * Copyright (C) 2016  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.dialogs

import android.app.Dialog
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment

import com.afollestad.materialdialogs.MaterialDialog

import java.io.Serializable

class GenericInputDialog : DialogFragment(), MaterialDialog.InputCallback {
    private lateinit var target: DialogListenerTarget
    private lateinit var dialogTag: String

    private val owner: GenericInputDialogListener?
        get() {
            return when (target) {
                DialogListenerTarget.ACTIVITY -> activity as GenericInputDialogListener?
                DialogListenerTarget.FRAGMENT -> targetFragment as GenericInputDialogListener?
                DialogListenerTarget.NONE -> null
            }
        }

    interface GenericInputDialogListener {
        fun onConfirmInput(tag: String, text: String)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = arguments
        val builder = args!!.getSerializable(ARG_BUILDER) as Builder
        target = args.getSerializable(ARG_TARGET) as DialogListenerTarget
        dialogTag = args.getString(ARG_TAG)

        val dialogBuilder = MaterialDialog.Builder(activity!!)

        var hint: CharSequence? = null
        var prefill: CharSequence? = null

        if (builder.title != null) {
            dialogBuilder.title(builder.title!!)
        } else if (builder.titleResId != 0) {
            dialogBuilder.title(builder.titleResId)
        }

        if (builder.message != null) {
            dialogBuilder.content(builder.message!!)
        } else if (builder.messageResId != 0) {
            dialogBuilder.content(builder.messageResId)
        }

        if (builder.positive != null) {
            dialogBuilder.positiveText(builder.positive!!)
        } else if (builder.positiveResId != 0) {
            dialogBuilder.positiveText(builder.positiveResId)
        }

        if (builder.negative != null) {
            dialogBuilder.negativeText(builder.negative!!)
        } else if (builder.negativeResId != 0) {
            dialogBuilder.negativeText(builder.negativeResId)
        }

        if (builder.hint != null) {
            hint = builder.hint
        } else if (builder.hintResId != 0) {
            hint = getText(builder.hintResId)
        }

        if (builder.preFill != null) {
            prefill = builder.preFill
        } else if (builder.preFillResId != 0) {
            prefill = getText(builder.preFillResId)
        }

        if (builder.inputType != 0) {
            dialogBuilder.inputType(builder.inputType)
        }

        dialogBuilder.input(hint, prefill, builder.allowEmpty, this)

        val dialog = dialogBuilder.build()

        isCancelable = false
        dialog.setCanceledOnTouchOutside(false)

        return dialog
    }

    override fun onInput(dialog: MaterialDialog, input: CharSequence) {
        owner?.onConfirmInput(dialogTag, input.toString())
    }

    class Builder : Serializable {
        internal var title: String? = null
        @StringRes
        internal var titleResId: Int = 0
        internal var message: String? = null
        @StringRes
        internal var messageResId: Int = 0
        internal var positive: String? = null
        @StringRes
        internal var positiveResId: Int = 0
        internal var negative: String? = null
        @StringRes
        internal var negativeResId: Int = 0
        internal var hint: String? = null
        @StringRes
        internal var hintResId: Int = 0
        internal var preFill: String? = null
        @StringRes
        internal var preFillResId: Int = 0
        internal var allowEmpty = true
        internal var inputType: Int = 0

        fun title(title: String?): Builder {
            this.title = title
            titleResId = 0
            return this
        }

        fun title(@StringRes titleResId: Int): Builder {
            title = null
            this.titleResId = titleResId
            return this
        }

        fun message(message: String?): Builder {
            this.message = message
            messageResId = 0
            return this
        }

        fun message(@StringRes messageResId: Int): Builder {
            message = null
            this.messageResId = messageResId
            return this
        }

        fun positive(text: String?): Builder {
            positive = text
            positiveResId = 0
            return this
        }

        fun positive(@StringRes textResId: Int): Builder {
            positive = null
            positiveResId = textResId
            return this
        }

        fun negative(text: String?): Builder {
            negative = text
            negativeResId = 0
            return this
        }

        fun negative(@StringRes textResId: Int): Builder {
            negative = null
            negativeResId = textResId
            return this
        }

        fun hint(text: String?): Builder {
            hint = text
            hintResId = 0
            return this
        }

        fun hint(@StringRes textResId: Int): Builder {
            hint = null
            hintResId = textResId
            return this
        }

        fun prefill(text: String?): Builder {
            preFill = text
            preFillResId = 0
            return this
        }

        fun prefill(@StringRes textResId: Int): Builder {
            preFill = null
            preFillResId = textResId
            return this
        }

        fun allowEmpty(allowEmpty: Boolean): Builder {
            this.allowEmpty = allowEmpty
            return this
        }

        fun inputType(type: Int): Builder {
            inputType = type
            return this
        }

        fun buildFromFragment(tag: String, parent: Fragment): GenericInputDialog {
            val args = Bundle()
            args.putSerializable(ARG_BUILDER, this)
            args.putSerializable(ARG_TARGET, DialogListenerTarget.FRAGMENT)
            args.putString(ARG_TAG, tag)

            val dialog = GenericInputDialog()
            dialog.setTargetFragment(parent, 0)
            dialog.arguments = args
            return dialog
        }

        fun buildFromActivity(tag: String): GenericInputDialog {
            val args = Bundle()
            args.putSerializable(ARG_BUILDER, this)
            args.putSerializable(ARG_TARGET, DialogListenerTarget.ACTIVITY)
            args.putString(ARG_TAG, tag)

            val dialog = GenericInputDialog()
            dialog.arguments = args
            return dialog
        }
    }

    companion object {
        private const val ARG_BUILDER = "builder"
        private const val ARG_TARGET = "target"
        private const val ARG_TAG = "tag"
    }
}